<a href="https://colab.research.google.com/github/Sepren/MachineLearning/blob/main/hw1_data_analysis_pandas_checkpoint.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <center> Домашнее задание № 1
## <center> Анализ данных по сердечно-сосудистым заболеваниям

**В задании предлагается с помощью Pandas ответить на несколько вопросов по данным о сердечно-сосудистых заболеваниях (в репозитории в папке data есть текстовик с ссылкой, выгрузите в эту же папку весь архив, чтобы код работал). Данные использовались в соревновании [Ml Boot Camp 5](http://mlbootcamp.ru/round/12/sandbox/).**

**Заполните код в клетках (где написано "Ваш код здесь") и ответьте на вопросы в [веб-форме](https://docs.google.com/forms/d/1NPMIal9jOQzBhy3KEc05rqUUMm228im3z5L2KZ-Jv_4/edit?usp=sharing).**

В соревновании предлагалось определить наличие/отсутствие сердечно-сосудистых заболеваний (ССЗ) по результатам осмотра пациента.

**Описание данных.**

Датасет сформирован из реальных клинических анализов, и в нём используются признаки, которые можно разбить на 3 группы:

Объективные признаки:

 - Возраст (age)
 - Рост (height)
 - Вес (weight)
 - Пол (gender)


Результаты измерения:

 - Артериальное давление верхнее и нижнее (ap_hi, ap_lo)
 - Холестерин (cholesterol)
 - Глюкоза (gluc)


Субъективные признаки (со слов пациентов):

 - Курение (smoke)
 - Употребление алкоголя (alco)
 - Физическая активность (active)

Целевой признак (который интересно будет прогнозировать):
 - Наличие сердечно-сосудистых заболеваний по результатам классического врачебного осмотра (cardio)

Значения показателей холестерина и глюкозы представлены одним из трех классов: норма, выше нормы, значительно выше нормы. Значения субъективных признаков — бинарны.

Все показатели даны на момент осмотра.

Мы будем работать только с обучающей выборкой и с помощью `Pandas` проведем первичный анализ данных.

Из библиотек нам понадобятся только `NumPy` и `Pandas`.

In [2]:
import numpy as np
import pandas as pd

Считываем данные из CSV-файла в объект pandas DataFrame.  

In [4]:
df = pd.read_csv('mlbootcamp5_train.csv', sep=';',
                 index_col='id')

Посмотрим на первые 5 записей.

In [5]:
df.head()

age  gender  height  weight  ap_hi  ap_lo  cholesterol  gluc  smoke  \
id                                                                          
0   18393       2     168    62.0    110     80            1     1      0   
1   20228       1     156    85.0    140     90            3     1      0   
2   18857       1     165    64.0    130     70            3     1      0   
3   17623       2     169    82.0    150    100            1     1      0   
4   17474       1     156    56.0    100     60            1     1      0   

    alco  active  cardio  
id                        
0      0       1       0  
1      0       1       1  
2      0       0       1  
3      0       1       1  
4      0       0       0

**<font color='red'>Вопрос 1 (1 балл).</font> Сколько мужчин и женщин представлено в этом наборе данных? Не было дано расшифровки признака "пол" (какому полу соответствует 1, а какому – 2 в признаке ``gender``) – это определите, посмотрев также на рост при разумном предположении, что в среднем мужчины выше (здесь и далее под средним понимается среднее арифметическое).**

**<font color='red'>Варианты:</font>**
- 45470 женщин и 24530 мужчин
- 45470 мужчин и 24530 женщин
- 45530 женщин и 24470 мужчин
- 45530 мужчин и 24470 женщин

In [8]:
df["gender"].value_counts()
# df.groupby('gender').mean()
# 2 - муж, 1 - жена

1    45530
2    24470
Name: gender, dtype: int64

**<font color='red'>Вопрос 2 (1 балл).</font> Кто в среднем реже указывает, что употребляет алкоголь – мужчины или женщины?**

**<font color='red'>Варианты:</font>**
- мужчины
- женщины

In [7]:
df.query('alco==0 and gender==2').shape[0] / df.query('gender==2').shape[0], df.query('alco==0 and gender==1').shape[0] / df.query('gender==1').shape[0]
# женщины

(0.8936248467511239, 0.9745003294531078)

**<font color='red'>Вопрос 3 (1 балл).</font> Во сколько раз (округленно, `round`) процент курящих среди мужчин больше, чем процент курящих среди женщин (по крайней мере, по этим анкетным данным)?**

**<font color='red'>Варианты:</font>**
- 4
- 8
- 12
- 16

In [ ]:
(df.query('smoke==1 and gender==2').shape[0] / df.query('gender==2').shape[0] * 100) / (df.query('smoke==1 and gender==1').shape[0] / df.query('gender==1').shape[0] * 100)
# 12

12.257833097333833

**<font color='red'>Вопрос 4 (1 балл).</font> Вы наверняка заметили, что значения возраста какие-то странные. Догадайтесь, в чём здесь измеряется возраст, и ответьте, на сколько месяцев (примерно) отличаются медианные значения возраста курящих и некурящих.**

**<font color='red'>Варианты:</font>**
- 10
- 15
- 20
- 25

In [ ]:
df.query('smoke==1').mean() * 12 - df.query('smoke==0').mean() * 12
# 15

age           -13.629495
gender          6.825181
height         65.333538
weight         41.303991
ap_hi          -6.011025
ap_lo          41.375615
cholesterol     0.298159
gluc           -0.115223
smoke          12.000000
alco            3.247346
active          0.434744
cardio         -0.327766
dtype: float64

**<font color='red'>Вопрос 5 (2 балла).</font> В [статье](https://ru.wikipedia.org/wiki/Сердечно-сосудистый_риск) на Википедии про сердечно-сосудистый риск показана шкала SCORE для расчёта риска смерти от сердечно-сосудистого заболевания в ближайшие 10 лет. Вот она:**
<img src='https://habrastorage.org/webt/gg/38/vf/gg38vfl5ql8obqjmlsvt9es76p8.png' width=70%>

Давайте посмотрим на правый верхний прямоугольник, отображающий сегмент курящих мужчин в возрасте от 60 до 64 лет включительно. (Неочевидно, но тут для возраста и давления цифры означают верхнюю границу, и она не включается).

Видим 9-ку в левом нижнем углу этого прямоугольника и 47 – в правом верхнем. То есть если при этом систолическое (т.е. верхнее) артериальное давление – меньше 120 мм рт.ст., а уровень холестерина – 4 ммоль/л, то риск ССЗ оценивается примерно в 5 раз ниже, чем если бы давление лежало в интервале [160, 180), а холестерина было бы 8 ммоль/л.

Давайте посчитаем аналогичное значение, но на наших данных.

Уточнения:
- Создайте новый признак ``age_years`` – возраст в годах, округлив до целых (`round`). Для данного примера отберите курящих мужчин от 60 до 64 лет включительно
- Категории уровня холестрина на рисунке и в наших данных отличаются. Отображение значений на картинке в значения признака ``cholesterol`` следующее: 4 ммоль/л $\rightarrow$ 1, 5-7 ммоль/л $\rightarrow$ 2, 8 ммоль/л $\rightarrow$ 3.
- Интересуют 2 подвыборки курящих мужчин возраста от 60 до 64 лет включительно: первая с верхним артериальным давлением строго меньше 120 мм рт.ст. и концентрацией холестерина – 4 ммоль/л, а вторая – с верхним артериальным давлением от 160 (включительно) до 180 мм рт.ст. (не включительно) и концентрацией холестерина – 8 ммоль/л.

Во сколько раз (округленно, `round`) отличаются доли больных людей (согласно целевому признаку, `cardio`) в этих двух подвыборках? Посчитайте на наших данных.

**<font color='red'>Варианты:</font>**
- 2
- 3
- 4
- 5

In [2]:
"""
df['age_years'] = (df['age'] / 365.25).round().astype(int)

filtered_df = df[(df['gender'] == male_gender) & (df['smoke'] == 1) & (df['age_years'] >= 60) & (df['age_years'] <= 64)]

subset1 = filtered_df[(filtered_df['ap_hi'] < 120) & (filtered_df['cholesterol'] == 1)]

subset2 = filtered_df[(filtered_df['ap_hi'] >= 160) & (filtered_df['ap_hi'] < 180) & (filtered_df['cholesterol'] == 3)]

ratio1 = subset1['cardio'].mean()
ratio2 = subset2['cardio'].mean()

difference_ratio = round(ratio2 / ratio1)
print(difference_ratio)
"""

"\ndf['age_years'] = (df['age'] / 365.25).round().astype(int)\n\nfiltered_df = df[(df['gender'] == male_gender) & (df['smoke'] == 1) & (df['age_years'] >= 60) & (df['age_years'] <= 64)]\n\nsubset1 = filtered_df[(filtered_df['ap_hi'] < 120) & (filtered_df['cholesterol'] == 1)]\n\nsubset2 = filtered_df[(filtered_df['ap_hi'] >= 160) & (filtered_df['ap_hi'] < 180) & (filtered_df['cholesterol'] == 3)]\n\nratio1 = subset1['cardio'].mean()\nratio2 = subset2['cardio'].mean()\n\ndifference_ratio = round(ratio2 / ratio1)\nprint(difference_ratio)\n"

**<font color='red'>Вопрос 6 (2 балла).</font> Постройте новый признак – BMI ([Body Mass Index](https://en.wikipedia.org/wiki/Body_mass_index)). Для этого надо вес в килограммах поделить на квадрат роста в метрах. Нормальными считаются значения BMI от 18.5 до 25. Выберите верные утверждения.**

**<font color='red'>Утверждения:</font>**
- Медианный BMI по выборке превышает норму
- У женщин в среднем BMI ниже, чем у мужчин
- У здоровых в среднем BMI выше, чем у больных
- В сегменте здоровых и непьющих мужчин в среднем BMI ближе к норме, чем в сегменте здоровых и непьющих женщин

In [11]:
df["BMI"] = df["weight"] / ((df["height"] / 100) ** 2)
df.BMI.mean()
df.query('gender==1').mean(), df.query('gender==2').mean()

"""
df['height_m'] = df['height'] / 100

df['BMI'] = df['weight'] / (df['height_m'] ** 2)

median_bmi = df['BMI'].median()

average_bmi_women = df[df['gender'] == female_gender]['BMI'].mean()
average_bmi_men = df[df['gender'] == male_gender]['BMI'].mean()

average_bmi_healthy = df[df['cardio'] == 0]['BMI'].mean()
average_bmi_sick = df[df['cardio'] == 1]['BMI'].mean()

average_bmi_healthy_non_alco_men = df[(df['cardio'] == 0) & (df['alco'] == 0) & (df['gender'] == male_gender)]['BMI'].mean()
average_bmi_healthy_non_alco_women = df[(df['cardio'] == 0) & (df['alco'] == 0) & (df['gender'] == female_gender)]['BMI'].mean()

print(f"Медианный BMI: {median_bmi}")
print(f"Средний BMI для женщин: {average_bmi_women}")
print(f"Средний BMI для мужчин: {average_bmi_men}")
print(f"Средний BMI для здоровых: {average_bmi_healthy}")
print(f"Средний BMI для больных: {average_bmi_sick}")
print(f"Средний BMI для здоровых и непьющих мужчин: {average_bmi_healthy_non_alco_men}")
print(f"Средний BMI для здоровых и непьющих женщин: {average_bmi_healthy_non_alco_women}")
"""

(age            19510.124577
 gender             1.000000
 height           161.355612
 weight            72.565605
 ap_hi            128.139249
 ap_lo             94.522776
 cholesterol        1.384735
 gluc               1.235054
 smoke              0.017856
 alco               0.025500
 active             0.802021
 cardio             0.496727
 BMI               27.987583
 dtype: float64,
 age            19392.097875
 gender             2.000000
 height           169.947895
 weight            77.257307
 ap_hi            130.078872
 ap_lo            100.551982
 cholesterol        1.333633
 gluc               1.210462
 smoke              0.218880
 alco               0.106375
 active             0.806906
 cardio             0.505231
 BMI               26.754442
 dtype: float64)

**<font color='red'>Вопрос 7 (2 балла).</font> ** Можно заметить, что данные не особо-то чистые, много в них всякой "грязи" и неточностей. Еще лучше мы это увидим, когда обсудим визуализацию данных.

Отфильтруйте следующие сегменты пациентов (считаем это ошибками в данных)
- указанное нижнее значение артериального давления строго выше верхнего
- рост строго меньше 2.5%-перцентили или строго больше 97.5%-перцентили (используйте `pd.Series.quantile`, если не знаете, что это такое – прочитайте)
- вес строго меньше 2.5%-перцентили или строго больше 97.5%-перцентили

Это вовсе не вся чистка данных, которую можно было проделать, но пока остановимся на этом.

Сколько процентов данных (округленно, `round`) мы выбросили?

**<font color='red'>Варианты:</font>**
- 8
- 9
- 10
- 11

In [ ]:
"""
initial_count = len(df)
df = df[df['ap_lo'] <= df['ap_hi']]
lower_bound_height = df['height'].quantile(0.025)
upper_bound_height = df['height'].quantile(0.975)
df = df[(df['height'] >= lower_bound_height) & (df['height'] <= upper_bound_height)]
lower_bound_weight = df['weight'].quantile(0.025)
upper_bound_weight = df['weight'].quantile(0.975)
df = df[(df['weight'] >= lower_bound_weight) & (df['weight'] <= upper_bound_weight)]
final_count = len(df)
filtered_percentage = round((initial_count - final_count) / initial_count * 100)
print(f"Процент отфильтрованных данных: {filtered_percentage}%")
"""